src/distributions/uniform.rs

```diff
@@ -475,6 +475,139 @@ uniform_int_impl! { usize, isize, usize, isize, usize }
 #[cfg(feature = "i128_support")]
 uniform_int_impl! { u128, u128, u128, i128, u128 }
 
+#[cfg(feature = "simd_support")]
+macro_rules! uniform_simd_int_impl {
+    ($ty:ident, $unsigned:ident, $signed:ident, $u_scalar:ident) => {
+        // The "pick the largest zone that can fit in an `u32`" optimization
+        // is less useful here. Multiple lanes complicate things, we don't
+        // know the PRNG's minimal output size, and casting to a larger vector
+        // is generally a bad idea for SIMD performance. The user can still
+        // implement it manually.
+
+        // TODO: look into `Uniform::<u32x4>::new(0u32, 100)` functionality
+        //       perhaps `impl SampleUniform for $u_scalar`?
+        impl SampleUniform for $ty {
+            type Sampler = UniformInt<$ty>;
+        }
+
+        impl UniformSampler for UniformInt<$ty> {
+            type X = $ty;
+
+            #[inline] // if the range is constant, this helps LLVM to do the
+                      // calculations at compile-time.
+                      // TODO: test this ^ for SIMD
+            fn new<B1, B2>(low_b: B1, high_b: B2) -> Self
+                where B1: SampleBorrow<Self::X> + Sized,
+                      B2: SampleBorrow<Self::X> + Sized
+            {
+                let low = *low_b.borrow();
+                let high = *high_b.borrow();
+                assert!(low.lt(high).all(), "Uniform::new called with `low >= high`");
+                UniformSampler::new_inclusive(low, high - 1)
+            }
+
+            #[inline] // if the range is constant, this helps LLVM to do the
+                      // calculations at compile-time.
+            fn new_inclusive<B1, B2>(low_b: B1, high_b: B2) -> Self
+                where B1: SampleBorrow<Self::X> + Sized,
+                      B2: SampleBorrow<Self::X> + Sized
+            {
+                let low = *low_b.borrow();
+                let high = *high_b.borrow();
+                assert!(low.le(high).all(),
+                        "Uniform::new_inclusive called with `low > high`");
+                let unsigned_max = ::core::$u_scalar::MAX;
+
+                let range = $unsigned::from((high - low) + 1);
+                // The `range == 0` case is complicated with multiple lanes,
+                // avoid it.
+                assert!(range.eq($unsigned::splat(0)).none(),
+                        "Uniform::new_inclusive called with the full range");
```

Comment:
One way to deal actually implement this would be to set zone to whatever here for lanes where range==0. Then in `sample`, change the return statement to:

```rust
return range.ne($unsigned::splat(0)).select(self.low + $ty::from(hi), v);
```

It obviously adds some overhead, but so does our handling in the non-simd case.

It's possible that the overhead is bigger here though. In particular I'm not sure if the `$unsigned::splat(0)` call will get compiled into something fast. It would for non-simd types, but I don't know SIMD instructions well enough to know if that's the case here.

If there's no way to make the splat operation fast, you could also try:

```rust
return range.ne(hi).select(self.low + $ty::from(hi), v);
```